dp_check/dp_check.go

```diff
@@ -142,32 +235,103 @@ func main() {
 	}
-	// Check IPv6
-	runCheck("IPv6 routes", func() error {
-		var out string
-		var err error
-		cmd := "ip -6 route show"
-		infoLog.Printf("Check IPv6 routes with:|%v|...\n", cmd)
-		if out, err = getOutput(cmd); err != nil {
+
+	// Check if dp_check is running on a VM
+	runCheck("Running on GCE VM", func() error {
+		ret, err := isRunningOnGCP()
+		if err != nil {
 			return err
 		}
-		if !strings.Contains(out, "2001:4860:8040::/42") {
-			return fmt.Errorf(`Missing route prefix to backends: 2001:4860:8040::/42.
-IPv6 route setup either failed or hasn't been attempted`)
+		if !ret {
+			return fmt.Errorf("dp_check is not running on a GCE VM, this tool will not work as intended")
 		}
 		return nil
 	})
-	runCheck("IPv6 addresses", func() error {
-		var out string
-		var err error
-		cmd := "ip -6 addr show"
-		infoLog.Printf("Check IPv6 addresses with:|%v|...\n", cmd)
-		if out, err = getOutput(cmd); err != nil {
+
+	// Check connection to metadata server
+	runCheck("DirectPath availability", func() error {
+		client := &http.Client{}
+		metadataServerUrl := "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s"
+		req, err := http.NewRequest("GET", metadataServerUrl, nil)
+		if err != nil {
+			return err
+		}
+		req.Header.Add("Metadata-Flavor", "Google")
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
 			return err
 		}
-		if !strings.Contains(out, "inet6 2600") {
+		ipv6FromMetadataServer = net.ParseIP(strings.TrimSuffix(string(body), "\n"))
+
+		infoLog.Printf("Check if this VM enables DirectPath by sending http GET request to %s", metadataServerUrl)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode == 200 {
+			return nil
+		}
+		if resp.StatusCode == 404 {
+			return fmt.Errorf("This VM doesn't have DirectPath access")
+		}
+		return fmt.Errorf("Connecting to metadata server failed. There may be a bug that may be causing a larger outage")
+	})
+
+	runCheck("IPv6 addresses and routes", func() error {
+		ifaces, err := net.Interfaces()
+		if err != nil {
+			return err
+		}
+		// Go through all interfaces on the VM to see if IPv6 is enabled and if there is an IPv6 address, then check against the IPv6 address returned from metadataserver
+		var directPathNetworkInterface net.Interface
+		var ipv6Enabled bool
+		for _, iface := range ifaces {
+			if iface.Flags&net.FlagLoopback != 0 {
+				continue
+			}
+			if iface.Flags&net.FlagUp != net.FlagUp {
+				continue
+			}
+			ifaddrs, err := iface.Addrs()
+			if err != nil {
+				return err
+			}
+			for _, ifaddr := range ifaddrs {
+				ip := ifaddr.(*net.IPNet).IP
+				if ip.To4() != nil {
+					continue
+				}
+				ipv6Enabled = true
+				if ip.Equal(ipv6FromMetadataServer) {
+					directPathNetworkInterface = iface
+					break
+				}
+			}
+			if directPathNetworkInterface.Name != "" {
+				break
+			}
+		}
+		if !ipv6Enabled {
```

Comment:
I don't think the separate `ipv6Enabled` check is necessary to have on top of the check for `directpathNetworkInterface`, because I'm pretty sure that we'll set `ipv6Enabled = true` even in many cases in which IPv6 hasn't been properly set up (because interfaces will still have locally scoped ipv6 addresses).

Can we remove this check, but move the `IPv6 DHCP setup either failed or hasn't been attempted` error message down into the error message for `if directPathNetworkInterface.Name == "" {` ?